components/scream/src/share/tests/coarsening_remapper_tests.cpp

```diff
@@ -0,0 +1,417 @@
+  }
+
+  grid_ptr_type get_ov_tgt_grid () const {
+    return m_ov_tgt_grid;
+  }
+
+  int gid2lid (const gid_t gid, const grid_ptr_type& grid) const {
+    return CoarseningRemapper::gid2lid(gid,grid);
+  }
+};
+
+template<typename ViewT>
+typename ViewT::HostMirror
+cmvc (const ViewT& v) {
+  auto vh = Kokkos::create_mirror_view(v);
+  Kokkos::deep_copy(vh,v);
+  return vh;
+}
+
+template<typename ViewT>
+bool contains (const ViewT& v, const typename ViewT::traits::value_type& entry) {
+  const auto vh = cmvc (v);
+  const auto beg = vh.data();
+  const auto end = vh.data() + vh.size();
+  for (auto it=beg; it!=end; ++it) {
+    if (*it == entry) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void print (const std::string& msg, const ekat::Comm& comm) {
+  if (comm.am_i_root()) {
+    printf("%s",msg.c_str());
+  }
+}
+
+TEST_CASE ("coarsening_remap") {
+
+  // -------------------------------------- //
+  //           Init MPI and PIO             //
+  // -------------------------------------- //
+
+  ekat::Comm comm(MPI_COMM_WORLD);
+
+  MPI_Fint fcomm = MPI_Comm_c2f(comm.mpi_comm());
+  scorpio::eam_init_pio_subsystem(fcomm);
+
+  // -------------------------------------- //
+  //           Set grid/map sizes           //
+  // -------------------------------------- //
+
+  const int nldofs_src = 10;
+  const int nldofs_tgt =  5;
+  const int ngdofs_src = nldofs_src*comm.size();
+  const int ngdofs_tgt = nldofs_tgt*comm.size();
+  const int nnz = ngdofs_src;
+
+  // -------------------------------------- //
+  //           Create a map file            //
+  // -------------------------------------- //
+
+  print (" -> creating map file ...\n",comm);
+
+  std::string filename = "coarsening_map_file_np" + std::to_string(comm.size()) + ".nc";
+  scorpio::register_file(filename, scorpio::FileMode::Write);
+
+  scorpio::register_dimension(filename,"n_a", "n_a", ngdofs_src);
+  scorpio::register_dimension(filename,"n_b", "n_b", ngdofs_tgt);
+  scorpio::register_dimension(filename,"n_s", "n_s", nnz);
+
+  scorpio::register_variable(filename,"col","col","none",{"n_s"},"real","int","Real-nnz");
+  scorpio::register_variable(filename,"row","row","none",{"n_s"},"real","int","Real-nnz");
+  scorpio::register_variable(filename,"S","S","none",{"n_s"},"real","real","Real-nnz");
+
+  std::vector<std::int64_t> dofs (nldofs_src);
+  std::iota(dofs.begin(),dofs.end(),comm.rank()*nldofs_src);
+  scorpio::set_dof(filename,"col",nldofs_src,dofs.data());
+  scorpio::set_dof(filename,"row",nldofs_src,dofs.data());
+  scorpio::set_dof(filename,"S",nldofs_src,dofs.data());
+  
+  scorpio::eam_pio_enddef(filename);
+
+  // Create triplets: tgt entry K is the avg of src entries K and K+ngdofs_tgt
+  // NOTE: add 1 to row/col indices, since e3sm map files indices are 1-based
+  std::vector<Real> col,row,S;
+  for (int i=0; i<nldofs_tgt; ++i) {
+    row.push_back(1+i+nldofs_tgt*comm.rank());
+    col.push_back(1+i+nldofs_tgt*comm.rank());
+    S.push_back(0.5);
```

Comment:
Here you are essentially making the weights 0.5 everywhere which doesn't necessarily check what happens when they are not uniform.  We could make this value 0.7 and the next one 0.3, or any combo that adds to 1.0.  Then if there is a mismatch in weight order we might catch it.  You could even have a random real between 0 and 1 and make this one `wgt` and the other `1-wgt`.

Reply:
I picked 0.5 b/c multiplying by powers of two is an exact op in floating point arithmetic (so long as we don't over/underflow, it's just an exponent shift). I suppose multiplying by 0.25 is too, and the counterpart 0.75 should be fine too.

Reply:
But to be precise, I want to avoid weights that do not have an exact FP representation, since I want the checks to be exact, rather than require some tolerance.